example/resnet/resnet.go

```diff
@@ -143,14 +112,66 @@ func main() {
 
 	model := models.Resnet50()
 	model.To(device)
+	model.Train(true)
 
 	optimizer := torch.SGD(lr, momentum, 0, weightDecay, false)
 	optimizer.AddParameters(model.Parameters())
 
-	start := time.Now()
-	for epoch := 0; epoch < epochs; epoch++ {
-		adjustLearningRate(optimizer, epoch, lr)
-		train(model, optimizer, batchSize, device, *trainFile)
+	f, e := os.Open(*trainFile)
+	if e != nil {
+		log.Fatal(e)
+	}
+	// build label vocabulary
+	vocab, e := datasets.BuildLabelVocabulary(f)
+	if e != nil {
+		log.Fatal(e)
+	}
+	log.Print("building label vocabulary done.")
+
+	batchs := itersEachEpoch(trainSamples, batchSize)
+	batch := 0
+	epoch := 0
+	adjustLearningRate(optimizer, epoch, lr)
+	skipSamples := 0
+	startTime := time.Now()
+	for epoch < epochs {
+		// seek to 0 of the file reader, and create an ImageNet loader
+		if _, e := f.Seek(0, io.SeekStart); e != nil {
+			log.Fatal(e)
+		}
+		loader, e := imageNetLoader(f, vocab, batchSize, skipSamples)
+		if e != nil {
+			panic(e)
+		}
+		for loader.Scan() {
+			batch++
+			image, target := loader.Minibatch()
+			loss, acc1, acc5 := trainOneBatch(image.To(device, torch.Float), target.To(device, torch.Long), model, optimizer)
+			if batch%logInterval == 0 {
+				throughput := float64(batch/logInterval) / time.Since(startTime).Seconds()
+				log.Printf("Epoch: %d, Batch: %d, loss:%f, acc1: %f, acc5:%f, throughput: %f samples/secs",
+					epoch, batch, loss, acc1, acc5, throughput)
+				startTime = time.Now()
+			}
+			if batch == batchs {
+				break
+			}
+		}
+		if batch == batchs {
+			// go to next epoch
+			epoch++
+			adjustLearningRate(optimizer, epoch, lr)
+			batch = 0
+			skipSamples = rand.Intn(batchSize)
+			log.Printf("skip %d samples at the next epoch", skipSamples)
+		}
+	}
+}
+
+func itersEachEpoch(samples, batchSize int) int {
+	itersEachEpoch := trainSamples / batchSize
```

Comment:
These four lines are equivalent to one line:

```go
return (trainSamples + batchSize - 1) / batchSize
```

I learned this trick when I was 14 working on computer graphics. Later, I realized this is a very commonly-used trick https://stackoverflow.com/a/2745086 that we don't even need to add a comment to explain it.